### PR TITLE
TIP-706: Add ES PQB completeness sorter

### PR DESCRIPTION
**Pull request by damien-carcel**
## Description

This PR adds a PQB completeness sorter using Elasticsearch.

## Checklist

- [x] Add sorter integration tests for the field in the PQB
- [ ] Add missing integration to the PimCatalogXIntegration tests for sort
- [x] Add the sorter + specs
- [x] Update the reference documentation

## Definition Of Done

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added Behats                      | -
| Added integration tests           | OK
| Changelog updated                 | -
| Review and 2 GTM                  | OK
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/CompletenessSorter.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+     * Check if channel and value are valid
+     *
+     * @param string $locale
+     * @param string $scope
+     *
+     * @throws InvalidPropertyException
+     */
+    protected function checkLocaleAndScope($locale, $scope)
+    {
+        if (null === $locale) {
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/CompletenessSorter.php]: we should add a check if locale & channel are valid

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/CompletenessSorter.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    /**
+     * Check if channel and value are valid
+     *
+     * @param string $locale
+     * @param string $scope
+     *
+     * @throws InvalidPropertyException
+     */
+    protected function checkLocaleAndScope($locale, $scope)
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/CompletenessSorter.php]: $channel everywhere please

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/BaseFieldSorter.php [hunk 3]
```diff
@@ -57,7 +57,7 @@ public function addFieldSorter($field, $direction, $locale = null, $scope = null
             case Directions::ASCENDING:
                 $sortClause = [
                     $field => [
-                        'order' => 'ASC',
+                        'order'   => 'ASC',
```
Comment (SamirBoulil) [inline src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/BaseFieldSorter.php:60]: :)

**Review by SamirBoulil — APPROVED** [review]
Nice one ^^ GTM

File: src/Pim/Bundle/CatalogBundle/tests/integration/PQB/Filter/Metric/MetricFilterIntegration.php [hunk 4]
```diff
@@ -130,6 +130,7 @@ public function testOperatorSuperiorOrEquals()
         $this->assert($result, ['product_one', 'product_two']);
 
         $result = $this->executeFilter([['a_metric', Operators::GREATER_OR_EQUAL_THAN, ['amount' => 15000, 'unit' => 'WATT']]]);
+        $this->assert($result, ['product_two']);
```
Comment (skeleton) [inline src/Pim/Bundle/CatalogBundle/tests/integration/PQB/Filter/Metric/MetricFilterIntegration.php:133]: yeah, it's better with an assert ;)

**Review by skeleton — APPROVED** [review]

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/CompletenessSorter.php (hunk 2 above)
Comment (damien-carcel) [reply src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/CompletenessSorter.php]: What bothers me here is that the interface uses scope:
`public function addFieldSorter($field, $direction, $locale = null, $scope = null);`
and even the PQB itsels in its options:
`$pqb->addSorter('completeness', 'DESC', ['locale' => 'fr_FR', 'scope' => 'mobile']);`
So I though it was more consistent

Comment (damien-carcel) [reply src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/CompletenessSorter.php]: Will it be reworked?

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/CompletenessSorter.php (hunk 1 above)
Comment (damien-carcel) [reply src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/CompletenessSorter.php]: Wasn't sure about it. Fact is, we should do it everywhere, not just here.

File: src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/CompletenessSorter.php (hunk 2 above)
Comment (skeleton) [reply src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/CompletenessSorter.php]: you can change the variable in the interface, it's not a BC.
about `$pqb->addSorter('completeness', 'DESC', ['locale' => 'fr_FR', 'scope' => 'mobile'])`, I don't know, but you can keep like that, when I said "channel everywhere", I talked about your variable

Comment (damien-carcel) [reply src/Pim/Bundle/CatalogBundle/Elasticsearch/Sorter/CompletenessSorter.php]: OK then!

**Pull request merged**